### PR TITLE
Add CharShift plugin

### PR DESCRIPTION
**Pull request by gedankenexperimenter**
This plugin allows independent assignment of symbols to the unshifted and shifted positions of a key.  This is very similar to TopsTurvy and ShapeShifter, but more flexible than both, even combined. For example, it is possible to define the following keys (base/shifted):

- `,` / `;`
- `.` / `:`
- `(` / `[`
- `<` / `>`

Some of these are not currently possible, even with both TopsyTurvy and ShapeShifter.

Notes:
1. I'm open to changing the names `Unshifter` and `Unkey`.
2. I'm trying to set up structures that will make it possible to define the `Unkey` array in EEPROM, accessible by Chrysalis. I may need some help from @algernon to get that working, though AutoShift has perhaps taught me enough.
3. ~~I haven't finished commenting and documenting the code, so I don't consider it ready to merge yet, but that job won't take me very much longer.~~
4. This is the plugin that I consider essential that Kaleidoscope has been lacking, which is why I'm submitting it for inclusion in the core, rather than leaving it as a third-party plugin.  It's the first plugin that I wrote for Kaleidoglyph, and it's a great example of something that would have been a nightmare to implement before the `KeyEvent` changes, but which is relatively simple and straightforward now.

### CODE REVIEW

**Comment by gedankenexperimenter** [comment]
Curiously, if I convert everything to (`static`) class member variables and functions, the PROGMEM footprint drops by 24 bytes, and the RAM usage goes up by 1.  With just the boolean `reverse_shift_state_` as a (non-`static`) member variable, I get back the 1 byte of RAM, but PROGMEM usage goes up by 10 bytes.  I have only the most tenuous guesses why.

**Comment by gedankenexperimenter** [comment]
This should wait until after #1041 gets merged, but it's otherwise ready to go, unless someone can think of a better name for the plugin.

**Comment by obra** [comment]
`Unshift` makes sense. Looking at our current naming scheme for plugins,  the `er` part of `Unshifter` doesn't feel *quite* right, but I wouldn't refuse the plugin for that.

 `Unkey` isn't super evocative to me. `UnshiftedKey` or `Unshifted_key` feel like they might be clearer, at the expense of a few more keystrokes and have the advantage of sharing their names with the name of the plugin.

**Comment by gedankenexperimenter** [comment]
I think I named it "Unshifter" to imitate the existing "ShapeShifter".  "Unshift" doesn't sound any worse to be, but it also doesn't sound great.  I don't mind changing it.  I'm rather fond of "Unkey", but I agree that it's not particularly evocative of what it does. "DualKey" or something like that would be more descriptive, but would be too easily confused with DualUse.

If the plugin's name becomes "Unshift", then "UnshiftKey" would mean a `Key` whose value belongs to the "Unshift" plugin (i.e. `isUnshiftKey(key)`).  "UnshiftedKey" is perhaps too similar to that for my taste, and also might be a bit misleading, because the "unshifting" only happens for keys whose "upper" value doesn't have the `shift` flag. Hmm…

For a totally wild alternative, I'll throw out a different name idea for the plugin: Janus (or Janushift, perhaps), for the two-faced Roman god of boundaries.  Probably too esoteric, though.

**Comment by gedankenexperimenter** [comment]
Other name ideas:
- TransmogriShift (or Transhift or TransShift?)
- DoubleShift (or DualShift)
- TwoFace (after the Batman villain)
- Shiftling (a play on "changeling")
- DeceptiKeys (after the Transformers villains)

I'm just searching for inspiration; I'm under no illusions that any of these is actually great. =)

**Comment by gedankenexperimenter** [comment]
Here's another one, that's perhaps better: 
- ReShift

**Comment by obra** [comment]
Heh. I'm trying to get us away from 'clever' names for core plugins and more toward utilitarian names that are clear and unambiguous.

This feels like the kind of functionality that should be...invisible, because it's just always there.

Hrm. What would it look like if this were just baked into the core and available in keymaps without special setup?

**Comment by gedankenexperimenter** [comment]
Taking one of the `*Shift` options as an example, the plugin could be called `DoubleShift`, its `Key` test could be `isDoubleShiftKey()`, and it's config objects could be named `DoubleKey`s.  I think this pattern is good, as long as the prefix conveys the sense that there are two values, or that the value of the key can change.

**Comment by gedankenexperimenter** [comment]
> Heh. I'm trying to get us away from 'clever' names for core plugins and more toward utilitarian names that are clear and unambiguous.

Awww! And I just thought of calling it "WereKeys"! 🙄

> Hrm. What would it look like if this were just baked into the core and available in keymaps without special setup?

A couple of thoughts on that: 

1. Despite my personal feeling that this is essential to being able to fully define a custom keymap, this functionality doesn't seem to be available in QMK, which suggests that there's not much demand for it (though I do recall a story from Dr Bose about the development of HiFi where test subjects turned down the (unlabeled) bass and treble knobs consistently to make the music sound "right"…). 
2. I don't see a way to reasonably encode the data such that it can be included in the keymap without severe limitations on its flexibility.  If it were really obvious which symbols should be shifted and which shouldn't, and which ones should share a key, I think our "standard" keyboards would be a lot better than they are.  It could be done by doubling the number of layers stored, but that would be tremendously wasteful for just a handful of keys that get changed when shift is held.

I might just not be creative enough to think of real design alternatives for this one; it's existed in basically this form in my head since the beginning, about three years ago.

**Comment by gedankenexperimenter** [comment]
This plugin resolves #828 — it doesn't add the feature to ShapeShifter, but it does provide it via a new plugin.

**Comment by gedankenexperimenter** [comment]
How about the name `CharShift`?  Meant to imply that "shift with this key produces a different character than normal".

**Review by algernon — APPROVED** [review]
I think this plugin is fantastic, and we should definitely merge it ASAP. I did leave one comment in-line, which isn't too important right now, and shouldn't be a blocker either (hence my approval, rather than requesting change), but something we should figure out soonish.

As for naming... I'm responsible for most of the whimsical names, and I like them more than utilitarian names myself, but I also understand @obra's reasons for wanting more descriptive, rather than funny names. I'll let you two figure out what's the best name, is what I'm saying. :)

**Comment by algernon** [comment]
A few other comments: I agree with @gedankenexperimenter that it's not really feasible to encode what this plugin does onto the keymap itself, without severe limitations, and that'd render most of it's flexibility useless. So it's amazing to have this as a plugin, but I don't see how it could be a core keymap feature.

Regarding EEPROM storage: what if we do something similar to what we did with `DynamicMacros` and `DynamicTapDance`? Have a function to reserve space, something like `UnshifterConfiguration.reserveSpaceforUnkeys(8)`, which would reserve space for 8 pairs, thus 16 Keys, 32 bytes. I'd go with reserving space based on the amount of pairs, so we don't accidentally end up in a situation where we have useless padding bytes reserved. Once reserved, just put pairs there, and done. Since it's always pairs, we can random-access them, without using a sentinel. We also know the size, because we store that at reserve time.

The Focus parts could just dump / replace the reserved space a'la DynamicMacros, and Chrysalis would do the heavy lifting of interpreting it - which should be quite trivial, as its only pairs of Keys.

**Comment by obra** [comment]
charshift works for me

File: plugins/Kaleidoscope-Unshifter/src/kaleidoscope/plugin/Unshifter.cpp
Comment (gedankenexperimenter) [reply plugins/Kaleidoscope-Unshifter/src/kaleidoscope/plugin/Unshifter.cpp:126]: I'm 100% with you on this.  My trouble is not quite knowing how to do it.  Qukeys has the same problem, and I really want both plugins to be Chrysalis-configurable, once we figure out a standard way to do it.

**Comment by gedankenexperimenter** [comment]
I'll update this with the new name (CharShift) when I get a chance, and we'll see if it still looks good.

**Comment by gedankenexperimenter** [comment]
I have changed the plugin's name to `CharShift`, as that seems to be fairly descriptive and generally agreeable.
- The enigmatically-named `Unkey` object is now `CharShift::KeyPair`.
- The keymap markup has been changed from `UK(n)` to `CS(n)`.
- The helper macro for defining the progmem array is now `CS_KEYS()`.
- The `kaleidoscope::ranges` entries are now named `CS_FIRST` and `CS_LAST` to be consistent with other plugin ranges.

File: plugins/Kaleidoscope-Unshifter/src/kaleidoscope/plugin/Unshifter.cpp
Comment (algernon) [reply plugins/Kaleidoscope-Unshifter/src/kaleidoscope/plugin/Unshifter.cpp]: I can help with that. Once this is merged, I'll throw something quick together, and we can iterate from there.

**Review by algernon — APPROVED** [review]

**Comment by obra** [comment]
CharShift::KeyPair is so much better than UnKey :)

**Pull request merged**
